### PR TITLE
Issue #783 Fix all remaining MyPy errors

### LINKED ISSUE #783: Fix MyPy errors

### PR DESCRIPTION
**Pull request by Manangka**
Fixes #783

This commit fixes all remaining MyPy errors in the mf6 module

The PR is split in several commits, each commit contains a certain file in which the mypy errors are fixes or exist of a type of error that crosses multiple files. I suggest reviewing this PR commit by commit.

# Checklist

- [x] Links to correct issue
- [ ] Update changelog, if changes affect users
- [x] PR title starts with ``Issue #nr``, e.g. ``Issue #737``
- [ ] Unit tests were added
- [ ] **If feature added**: Added/extended example

### CODE REVIEW

File: imod/tests/test_mf6/test_well_highlvl.py [hunk 1]
```diff
@@ -75,7 +75,11 @@ def test_write_well(tmp_path: Path, grid_data, grid_data_layered, reference_outp
     top = ones_like(active.sel(layer=1), dtype=np.float64)
     bottom = grid_data_layered(np.float64, -2.0, 10)
     write_context = WriteContext(tmp_path)
-    well.write("packagename", globaltimes, False, write_context, active, top, bottom, k)
+    mf6_pkg = well.to_mf6_pkg(
+        active, top, bottom, k, False, write_context.is_partitioned
+    )
+    mf6_pkg.write("packagename", globaltimes, write_context)
+    mf6_pkg.write("packagename", globaltimes, write_context)
```
Comment (luitjansl) [inline imod/tests/test_mf6/test_well_highlvl.py:82]: why twice?

**Review by luitjansl — APPROVED** [review]
looks good

File: imod/mf6/buy.py [hunk 2]
```diff
@@ -89,7 +90,7 @@ class Buoyancy(Package):
 
     _pkg_id = "buy"
     _template = Package._initialize_template(_pkg_id)
-    _metadata_dict = {}
+    _metadata_dict: Dict[str, VariableMetaData] = {}
```
Comment (JoerivanEngelen) [inline imod/mf6/buy.py:93]: I see this ``_metadata_dict`` attribute is implemented for MetaSWAP, but not at all applied for Modflow 6. I see it is possibly used in the private method ``_get_vars_to_check``, but this one is never called, so I think we can remove it! I guess we once removed it, but this one was leftover?

File: imod/mf6/hfb.py [hunk 3]
```diff
@@ -343,6 +345,8 @@ def to_mf6_pkg(
 
         Parameters
         ----------
+        validate: bool
+            Run validation before converting
```
Comment (JoerivanEngelen) [inline imod/mf6/hfb.py:349]: Please put the arguments in the docstring in the same order as in the actual code.

File: imod/mf6/ist.py [hunk 4]
```diff
@@ -191,11 +194,11 @@ def __init__(
         mobile_immobile_mass_transfer_rate,
         decay=None,
         decay_sorbed=None,
-        bulk_density=None,
+        bulk_density: Optional[xr.DataArray] = None,
         distribution_coefficient=None,
```
Comment (JoerivanEngelen) [inline imod/mf6/ist.py:198]: Shouldn't ``decay``, ``decay_sorbed``, ``distribution_coefficient`` also get a type annotation?

File: imod/mf6/boundary_condition.py [hunk 5]
```diff
@@ -230,7 +232,12 @@ def _write_perioddata(self, directory, pkgname, binary):
             path = directory / pkgname / f"{self._pkg_id}.{ext}"
             self._write_datafile(path, bin_ds, binary=binary)
 
-    def write(self, pkgname: str, globaltimes: np.ndarray, write_context: WriteContext):
+    def write(
+        self,
+        pkgname: str,
+        globaltimes: Union[List, np.ndarray],
```
[the remark this answers is not in the document]
Comment (Manangka) [reply imod/mf6/boundary_condition.py:238]: Done

File: imod/mf6/mf6_wel_adapter.py [hunk 6]
```diff
@@ -53,11 +54,13 @@ def __init__(
         rate,
         concentration=None,
         concentration_boundary_type="aux",
+        save_flows: Optional[bool] = None,
         validate: bool = True,
     ):
         super().__init__()
         self.dataset["cellid"] = cellid
         self.dataset["rate"] = rate
+        self.dataset["save_flows"] = save_flows
```
Comment (JoerivanEngelen) [inline imod/mf6/mf6_wel_adapter.py:63]: Nice, related to this issue I think: https://github.com/Deltares/imod-python/issues/652

File: imod/mf6/sfr.py
Comment (JoerivanEngelen) [inline imod/mf6/sfr.py:1]: I think we can remove this file, as it is not used at all. This only makes it appear as if we support the SFR package.

File: imod/mf6/hfb.py [hunk 7]
```diff
@@ -333,6 +334,7 @@ def to_mf6_pkg(
         top: GridDataArray,
         bottom: GridDataArray,
         k: GridDataArray,
+        validate: bool = False,
```
[the remark this answers is not in the document]
Comment (Manangka) [reply imod/mf6/hfb.py:337]: This is part of the wel fix. I realized that the wel package is written the same was as the hfb package. I therefor written the write method of the wel package in the same way as the hfb package.

Except for the validate method. The wel package had a validate flag that the hfb was missing. By adding it to the hfb package both packages have the same kind of behavior and can be threaten the same

File: imod/tests/test_mf6/test_well_highlvl.py (hunk 1 above)
Comment (Manangka) [reply imod/tests/test_mf6/test_well_highlvl.py:82]: Ow that is a mistake.
Removed it

**Review by JoerivanEngelen — COMMENTED** [review]
Looks good, some comments.

I see we use a mix of ``List`` as well as ``list``. I see ``List`` and ``Dict`` are on the list of deprecated aliases since python 3.9, so we can replace them with ``list`` and ``dict`` respectively.

https://docs.python.org/3/library/typing.html#deprecated-aliases

File: imod/mf6/ist.py (hunk 4 above)
Comment (Manangka) [reply imod/mf6/ist.py:198]: They don't trigger a mypy error. Perhaps when we set the MyPy settings to a stricter profile they will. But at the moment i consider that out of scope

File: imod/mf6/sfr.py
Comment (Manangka) [reply imod/mf6/sfr.py:1]: Removed

**Comment by Manangka** [comment]
> Looks good, some comments.
> 
> I see we use a mix of `List` as well as `list`. I see `List` and `Dict` are on the list of deprecated aliases since python 3.9, so we can replace them with `list` and `dict` respectively.
> 
> https://docs.python.org/3/library/typing.html#deprecated-aliases

Replaced List/Dict by list/dict. There are other deprecated types but i consider it out of scope (mainly because it doesn't cause a mypy error with the current settings).

**Pull request merged**